### PR TITLE
Allow strings and numbers as labels in cut

### PR DESCRIPTION
**Pull request by skleinbo**
Referring to https://discourse.julialang.org/t/creating-data-bins-with-numeric-labels-with-cut/81281/10?u=skleinbo

Inference seems to be fine. Tests pass on `Julia 1.7`. 

Three  things:
1. There is a test error on `1.6`, but that is due to `Base.return_types` not inferring correctly. The actual return type is as expected.  
That's not ideal. Should the test be taken out, somehow altered to pass on older Julia versions (don't know how tbh), or executed conditionally?
2. Technically, mixed labels are possible when passed as a `AbstractVector{<:SupportedTypes}`. E.g. just passing `[1,"2",3]::Vector{Any}` fails, while `Union{Int, String}[1,"2",3]` is fine. Should an effort be made to automatically do a conversion? Given the IMHO obscurity of mixed labels, should this even be mentioned in the docs? I say no on both points.
3. A label formatter must always return the same concrete type. That's a disparity to 2. and one could jump through the same hoops to allow for mixed labels, but is it worth it?
4. The current `cut` implicitly casts labels of any `AbstractString` to `String`. While not promised, someone downstream might rely on that conversion. 

Thanks!

### CODE REVIEW

File: test/15_extras.jl [hunk 1]
```diff
@@ -87,6 +87,27 @@ const ≅ = isequal
...
+    x = @inferred cut(Vector{Union{T, Int}}([1,2,3,4,5,6,7,8]), 10:-2:0; labels)
+    @test x == [0,"2","2",4,4,"6","6",8]
+    @test isa(x, CategoricalVector{Union{Int64, String, T}})
+    @test isordered(x)
+    @test levels(x) == [0,"2",4,"6",8]
+
+    labels = (from, to , i; leftclosed, rightclosed) -> (to+from)/2
+    x = @inferred cut(Vector{Union{T, Int}}([1,2,3,4,5,6,7,8]), 0:2:10; labels)
+    @test x == [1.0, 3.0, 3.0, 5.0, 5.0, 7.0, 7.0, 9.0]
+    @test isa(x, CategoricalVector{Union{Float64, T}})
```
Comment (skleinbo) [inline test/15_extras.jl:108]: This fails on Julia 1.6

File: test/15_extras.jl [hunk 2]
```diff
@@ -87,6 +87,27 @@ const ≅ = isequal
...
+
+    labels = Union{Int64,String}[0,"2",4,"6",8]
+    x = @inferred cut(Vector{Union{T, Int}}([1,2,3,4,5,6,7,8]), 10:-2:0; labels)
+    @test x == [0,"2","2",4,4,"6","6",8]
+    @test isa(x, CategoricalVector{Union{Int64, String, T}})
+    @test isordered(x)
+    @test levels(x) == [0,"2",4,"6",8]
+
+    labels = (from, to , i; leftclosed, rightclosed) -> (to+from)/2
+    x = @inferred cut(Vector{Union{T, Int}}([1,2,3,4,5,6,7,8]), 0:2:10; labels)
```
Comment (skleinbo) [inline test/15_extras.jl:106]: This fails on Julia 1.6

**Comment by bkamins** [comment]
Could you please add to the docstring (or comment here) what contract change you exactly propose, i.e. the exact rules what inputs are accepted and what outputs they should produce. Thank you!

**Comment by skleinbo** [comment]
I'm sorry! Guess I jumped the gun there a bit.

Let `SupportedTypes = Union{AbstractString, AbstractChar, Number}`. 

_Currently_: `cut` takes a `labels::Union[AbstractString, Function::AbstractString}` kwarg and returns `CategoricalArray` with levels of type `String` or `Union{String, Missing}` depending on whether the input array has missing values or `extend=missing` is given.

_Proposed_: Allow `cut` to take more general `labels::Union{AbstractVector{L}, Function::L} where L<:SupportedTypes` and return `CategoricalArray` with levels of type `L` or `Union{L, Missing}` depending on whether the input array has missing values or `extend=missing` is given.

It appears the limitation to String levels was introduced due to stability issues in the return type. As far as I can tell, those do not exist in more recent versions of Julia.

The change would mean closer feature parity with [`pandas.cut`](https://pandas.pydata.org/docs/reference/api/pandas.cut.html)

In light of that, my above comments hopefully make more sense.

There is one thing that could potentially break something downstream. I've amended the original comment to keep bullet points together.

**Comment by bkamins** [comment]
OK - so I understand the use-case is to e.g. assign a floating middle of the interval as a level, so that later user can work with it programmatically. Right?

**Comment by skleinbo** [comment]
Exactly. I saw the (linked above) topic on discourse and wondered why that wasn't supported. Maybe you know of a good reason why not.

**Comment by bkamins** [comment]
@nalimilan designed it, but I guess the reason was that cut values are naturally considered to be labels. However, I agree that sometimes it is useful to be able to work with them programmatically later.

File: test/15_extras.jl [hunk 3]
```diff
@@ -87,6 +87,27 @@ const ≅ = isequal
     @test isa(x, CategoricalMatrix{Union{String, T}})
     @test isordered(x)
     @test levels(x) == ["[-2.134, 3.0)", "[3.0, 12.5)"]
+
+    labels = 0:2:8
+    x = @inferred cut(Vector{Union{T, Int}}([1,2,3,4,5,6,7,8]), 0:2:10; labels)
+    @test x == [0,2,2,4,4,6,6,8]
+    @test isa(x, CategoricalVector{Union{Int64, T}})
```
Comment (nalimilan) [inline test/15_extras.jl:94]:
```suggestion
    @test isa(x, CategoricalVector{Union{Int, T}})
```

File: test/15_extras.jl [hunk 4]
```diff
@@ -87,6 +87,27 @@ const ≅ = isequal
...
+    @test isordered(x)
+    @test levels(x) == [0,2,4,6,8]
+
+    labels = Union{Int64,String}[0,"2",4,"6",8]
+    x = @inferred cut(Vector{Union{T, Int}}([1,2,3,4,5,6,7,8]), 10:-2:0; labels)
+    @test x == [0,"2","2",4,4,"6","6",8]
+    @test isa(x, CategoricalVector{Union{Int64, String, T}})
```
Comment (nalimilan) [inline test/15_extras.jl:101]:
```suggestion
    @test isa(x, CategoricalVector{Union{Int, String, T}})
```

File: test/15_extras.jl [hunk 5]
```diff
@@ -87,6 +87,27 @@ const ≅ = isequal
     @test isa(x, CategoricalMatrix{Union{String, T}})
     @test isordered(x)
     @test levels(x) == ["[-2.134, 3.0)", "[3.0, 12.5)"]
+
+    labels = 0:2:8
+    x = @inferred cut(Vector{Union{T, Int}}([1,2,3,4,5,6,7,8]), 0:2:10; labels)
```
Comment (nalimilan) [inline test/15_extras.jl:92]: To support Julia 1.0:
```suggestion
    x = @inferred cut(Vector{Union{T, Int}}([1,2,3,4,5,6,7,8]), 0:2:10, labels= labels)
```

File: test/15_extras.jl [hunk 6]
```diff
@@ -87,6 +87,27 @@ const ≅ = isequal
...
+    @test x == [0,2,2,4,4,6,6,8]
+    @test isa(x, CategoricalVector{Union{Int64, T}})
+    @test isordered(x)
+    @test levels(x) == [0,2,4,6,8]
+
+    labels = Union{Int64,String}[0,"2",4,"6",8]
+    x = @inferred cut(Vector{Union{T, Int}}([1,2,3,4,5,6,7,8]), 10:-2:0; labels)
```
Comment (nalimilan) [inline test/15_extras.jl:99]:
```suggestion
    x = @inferred cut(Vector{Union{T, Int}}([1,2,3,4,5,6,7,8]), 10:-2:0, labels=labels)
```

**Review by nalimilan — COMMENTED** [review]
Thanks! I think the points you list are OK. I just wonder whether we can preserve inferrability.

**Comment by bkamins** [comment]
Could you please add tests for typically problematic cases like: duplicate numeric labels, or having both `-0.0` and `0.0` as labels? Thank you!

File: src/extras.jl [hunk 7]
```diff
@@ -180,8 +189,11 @@ function _cut(x::AbstractArray{T, N}, breaks::AbstractVector,
...
-        levs = Vector{String}(undef, n-1)
-        for i in 1:n-2
+        levs = [labels(from[1], to[1], 1,
+            leftclosed=breaks[1] != breaks[2], rightclosed=false)]
+        resize!(levs, n-1)
+        _L = eltype(levs)
+        for i in 2:n-2
             levs[i] = labels(from[i], to[i], i,
                              leftclosed=breaks[i] != breaks[i+1], rightclosed=false)
         end
```
[the remark this answers is not in the document]
Comment (skleinbo) [reply src/extras.jl:198]: It reads very clunkily, and the resizing is certainly not optimal. But at least on 1.7 inference works, while it fails with a comprehension.  I will keep it close to your original implementation
```julia
firstlevel = labels(from[1], to[1], 1, leftclosed=breaks[1] != breaks[2], rightclosed=false)
levs = Vector{typeof(firstlevel)}(undef, n-1)
levs[begin] = firstlevel
for i in 2:n-2
    levs[i] = labels(from[i], to[i], i, leftclosed=breaks[i] != breaks[i+1], rightclosed=false)
end
levs[end] = labels(from[end], to[end], n-1,
                   leftclosed=breaks[end-1] != breaks[end],
                   rightclosed=coalesce(extend, false))
``` 
It seems crucial to typecast `levs` explicitly. I tried a `map` too, but that doesn't work either.

**Comment by skleinbo** [comment]
@nalimilan Thank you for the thorough review. I will push changes soon. I think I have tracked down an inference issue and it may just work on Julia 1.6. I'll have to write more tests first though.

@bkamins Certainly!

**Comment by skleinbo** [comment]
Concerning `-0.0, 0.0`. This is valid independent of `cut`
```
julia> CategoricalArray([-0.0, 0.0, 1.0]; levels=[-0.0, 0.0, 1.0])
3-element CategoricalArray{Float64,1,UInt32}:
 -0.0
 0.0
 1.0
```
Do you want `cut` to check for duplicates with `==` rather than `isequal` regardless?

**Comment by bkamins** [comment]
Yes, it is valid in CategoricalArrays.jl and expected. I think the same should be respected in this PR (i.e. labels should be unique w.r.t. `isequal`). I just want it tested.

The point is that `cat` uses `==` for values:
```
julia> cut([-0.0, 0.0], 2)
ERROR: ArgumentError: could not extend breaks as all values are equal: please specify at least two breaks manually
```
but of course this is an orthogonal issue (it just prompted me to cover this case in tests)

**Comment by skleinbo** [comment]
> The point is that `cat` uses `==` for values:

But only when used as `cut(x, n)`, because then values pass through `quantile` I guess?`
Anyway, orthogonal issue like you say.

@nalimilan Tests on Julia 1.0 failed, because of a `begin` in indexing, and on `x86` because of an `Int64` instead of `Int` in a test. Regarding the former, I find it difficult to remember what syntax was valid at which point. Wasn't there a tool to check for backward compatible syntax?

File: test/15_extras.jl [hunk 8]
```diff
@@ -183,6 +226,12 @@ end
                                    labels=["1", "2", "2", "3"])
     @test_throws ArgumentError cut(1:10, [1, 3, 3, 5, 5, 11], allowempty=true,
                                    labels=["1", "2", "3", "2", "4"])
+
+    @test_throws ArgumentError cut([1,2,3,4,5,6,7,8], 0:2:10, labels=[0,1,1,2,3])
+    @test_throws ArgumentError cut([1,2,3,4,5,6,7,8], [0,2,2,6,8,10], labels=[0,1,1,2,3], allowempty=true)
+
+    fmt = (from, to, i; leftclosed, rightclosed) -> (i%2==0 ? to : 0.0)
+    @test_throws ArgumentError cut([1,2,3,4,5,6,7,8], 0:2:10, labels=fmt)
```
Comment (nalimilan) [inline test/15_extras.jl:234]:
```suggestion
    @test_throws ArgumentError cut(1:8, 0:2:10, labels=[0, 1, 1, 2, 3])
    @test_throws ArgumentError cut(1:8, [0, 2, 2, 6, 8, 10], labels=[0, 1, 1, 2, 3], allowempty=true)

    fmt = (from, to, i; leftclosed, rightclosed) -> (i % 2 == 0 ? to : 0.0)
    @test_throws ArgumentError cut(1:8, 0:2:10, labels=fmt)
```

File: test/15_extras.jl [hunk 9]
```diff
@@ -199,6 +248,17 @@ end
     @test err.value.msg == "could not extend breaks as all values are missing: please specify at least two breaks manually"
 
     @test cut([missing], [1, 2], extend=true) ≅ [missing]
+
+    @test cut(Float64[-0.0, 0.0, 1.0, 2.0, 3.0, 4.0], [-0.0, 0.0, 3.0], labels=[-0.0, 0.0, 3.0], extend=true) ==
+        [-0.0, 0.0, 0.0, 0.0, 3.0, 3.0]
+end
+
+@testset "cut with extend=missing" begin
+    x = @inferred cut(Float64[-0.0, 0.0, 1.0, 2.0, 3.0, 4.0], [-0.0, 0.0, 3.0], labels=[-0.0, 0.0], extend=missing)
```
Comment (nalimilan) [inline test/15_extras.jl:257]:
```suggestion
    @test cut([-0.0, 0.0, 1.0, 2.0, 3.0, 4.0], [-0.0, 0.0, 3.0],
              labels=[-0.0, 0.0, 3.0], extend=true) ==
        [-0.0, 0.0, 0.0, 0.0, 3.0, 3.0]
end

@testset "cut with extend=missing" begin
    x = @inferred cut([-0.0, 0.0, 1.0, 2.0, 3.0, 4.0], [-0.0, 0.0, 3.0],
                      labels=[-0.0, 0.0], extend=missing)
```

File: test/15_extras.jl [hunk 10]
```diff
@@ -87,6 +87,23 @@ const ≅ = isequal
...
+    @test isa(x, CategoricalVector{Union{Int, T}})
+    @test isordered(x)
+    @test levels(x) == [0,2,4,6,8]
+
+    labels = Union{Int, String}[0,"2",4,"6",8]
+    x = @inferred cut(Vector{Union{T, Int}}([1,2,3,4,5,6,7,8]), 10:-2:0, labels=labels)
+    @test x == [0,"2","2",4,4,"6","6",8]
+    @test isa(x, CategoricalVector{Union{Int, String, T}})
+    @test isordered(x)
+    @test levels(x) == [0,"2",4,"6",8]
```
Comment (nalimilan) [inline test/15_extras.jl:103]:
```suggestion
    labels = 0:2:8
    x = @inferred cut(Vector{Union{T, Int}}(1:8), 0:2:10, labels=labels)
    @test x == [0,2,2,4,4,6,6,8]
    @test isa(x, CategoricalVector{Union{Int, T}})
    @test isordered(x)
    @test levels(x) == [0, 2, 4, 6, 8]

    labels = Union{Int, String}[0, "2", 4, "6", 8]
    x = @inferred cut(Vector{Union{T, Int}}(1:8), 10:-2:0, labels=labels)
    @test x == [0, "2", "2", 4, 4, "6", "6", 8]
    @test isa(x, CategoricalVector{Union{Int, String, T}})
    @test isordered(x)
    @test levels(x) == [0, "2", 4, "6", 8]
```

File: src/extras.jl [hunk 11]
```diff
@@ -89,6 +89,14 @@ julia> cut(-1:0.5:1, 2, labels=["A", "B"])
  "B"
  "B"
 
+ julia> cut(-1:0.5:1, 2, labels=[-0.5, +0.5])
+ 5-element CategoricalArray{Float64,1,UInt32}:
+  -0.5
+  -0.5
+  0.5
+  0.5
+  0.5
```
Comment (nalimilan) [inline src/extras.jl:98]: The doctests seems to be failing here. Maybe a whitespace issue?

File: src/extras.jl (hunk 7 above)
[the remark this answers is not in the document]
Comment (nalimilan) [reply src/extras.jl]: OK. That's weird, but I guess using the type of the first value is OK as we don't expect people to use multiple types anyway.

**Comment by nalimilan** [comment]
> @nalimilan Tests on Julia 1.0 failed, because of a `begin` in indexing, and on `x86` because of an `Int64` instead of `Int` in a test. Regarding the former, I find it difficult to remember what syntax was valid at which point. Wasn't there a tool to check for backward compatible syntax?

I'm not aware of any automated tool. Compat.jl is often useful but it doesn't support all new syntaxes.

File: src/extras.jl [hunk 12]
```diff
@@ -52,7 +52,7 @@ also accept them.
   in `x` fall outside of the breaks; when `true`, breaks are automatically added to include
   all values in `x`, and the upper bound is included in the last interval; when `missing`,
   values outside of the breaks generate `missing` entries.
-* `labels::Union{AbstractVector,Function}`: a vector of strings giving the names to use for
+* `labels::Union{AbstractVector, Function}`: a vector of strings or numbers giving the names to use for
```
[the remark this answers is not in the document]
Comment (skleinbo) [reply src/extras.jl:55]: Huh, I swear I changed that before...

File: src/extras.jl [hunk 13]
```diff
@@ -152,10 +160,11 @@ function _cut(x::AbstractArray{T, N}, breaks::AbstractVector,
             end
         end
         if !ismissing(min_x) && breaks[1] > min_x
-            breaks = [min_x; breaks]
+            # this typecast is needed on Julia<1.7 for stable inference
+            breaks = eltype(breaks)[min_x; breaks]
```
[the remark this answers is not in the document]
Comment (skleinbo) [reply src/extras.jl:165]: Your last proposal seems to work fine. I'm just running tests locally before pushing. Indeed, a Float to Int conversion due to my implementation was the reason for a failing doc test.

File: src/extras.jl (hunk 11 above)
Comment (skleinbo) [reply src/extras.jl]: Indeed!

**Comment by nalimilan** [comment]
Thanks!

**Comment by nalimilan** [comment]
https://github.com/JuliaRegistries/General/pull/60890

**Comment by skleinbo** [comment]
Thank you for patient reviews :)
